src/scenic/syntax/veneer.py

```diff
@@ -111,1000 +132,1341 @@
+        values['parentOrientation'] = region.orientation[pos]
+    return Specifier("In", props, values)
+
+def ContainedIn(region):
+    """The :grammar:`contained in <region>` specifier.
+
+    Specifies :prop:'position', :prop:`parentOrientation`, and :prop:'regionContainedIn',
+    with no dependencies.
+    """
+    region = toType(region, Region, 'specifier "contained in R" with R not a Region')
+    pos = Region.uniformPointIn(region)
+    props = {'position': 1, 'regionContainedIn': 1}
+    values = {'position': pos, 'regionContainedIn': region}
+    if alwaysProvidesOrientation(region):
+        props['parentOrientation'] = 3
+        values['parentOrientation'] = region.orientation[pos]
+    return Specifier("ContainedIn", props, values)
+
+
+def On(thing):
+    """The :specifier:`on {X}` specifier.
+
+    Specifies :prop:'position' and :prop:'parentOrientation', depending on
+    :prop:'onDirection', :prop:'baseOffset', :prop:'contactTolerance'.
+
+    Note that while :specifier:`on` can be used with `Region`, `Object` and `Vector`,
+    it cannot be used with a distribution containing anything other than `Region`.
+
+    May be used to modify an already specified :prop:'position' property if a compatible
+    specifier has already done so.
+
+    Allowed forms:
+        on <region>
+        on <object>
+        on <vector>
+    """
+    if isA(thing, Object):
+        target = toType(thing.onSurface, Region, 'Cannot coax occupiedSpace of Object to Region')
+    elif canCoerce(thing, Region):
+        target = toType(thing, Region, 'specifier "on R" with R not a Region')
+    else:
+        target = toType(thing, Vector, "Cannot coax target to Vector")
+
+    props = {'position': 1}
+
+    if isA(target, Region) and alwaysProvidesOrientation(target):
+        props['parentOrientation'] = 2
+
+    def helper(context):
+        # Pick position based on whether we are specifying or modifying
+        if hasattr(context, 'position'):
+            if not isA(target, Region):
+                raise ValueError('Cannot use modifying "on V" with V a vector.')
+
+            pos = findOnHelper(target, context.position, context.onDirection)
+        elif isA(target, Vector):
+            pos = target
+        else:
+            pos = Region.uniformPointIn(target)
+
+        values = {}
+
+        contactOffset = Vector(0,0,context.contactTolerance) - context.baseOffset
+
+        if 'parentOrientation' in props:
+            values['parentOrientation'] = target.orientation[pos]
+            contactOffset = contactOffset.rotatedBy(values['parentOrientation'])
+
+        values['position'] = (pos + contactOffset)
+
+        return values
+
+    return ModifyingSpecifier("On", props, DelayedArgument({'onDirection', 'baseOffset', 'contactTolerance'}, helper),\
+        modifiable_props={'position'})
+
+@distributionFunction
+def findOnHelper(region, pos, on_direction):
+    on_pos = region.findOn(pos, on_direction=on_direction)
```

Comment:
This method seems to be implemented only by `MeshRegion`. Better add some more tests.